tensor_parallel.py

```diff
@@ -0,0 +1,315 @@
+                layer.self_attn.v_proj.weight.data = split_v_tensors[i].to(f"{device}:{i}")
+                layer.self_attn.o_proj.weight.data = split_o_tensors[i].to(f"{device}:{i}")
+   
+        bind(layer.self_attn, forward, "forward")
+
+        # Manually adjust the number of heads after sharding the self attention modules.
+        # For Llama2 models, your should adjust the number of heads separately.
+        assert model.model.config.num_attention_heads % dist.get_world_size() == 0
+        layer.self_attn.num_heads = model.model.config.num_attention_heads // dist.get_world_size()
+        layer.self_attn.num_key_value_heads = model.model.config.num_key_value_heads // dist.get_world_size()
+        layer.self_attn.hidden_size = model.model.config.hidden_size // dist.get_world_size()
+        
+        # shard the MLP part
+        layer.mlp.tp= tp_size
+        gate_proj = layer.mlp.gate_proj.weight
+        up_proj = layer.mlp.up_proj.weight
+        down_proj = layer.mlp.down_proj.weight
+        row_split_dim =  gate_proj.shape[0] // tp_size
+        col_split_dim = down_proj.shape[-1] // tp_size
+        split_gate_tensors = torch.split(gate_proj, row_split_dim, dim = 0)
+        split_up_tensors = torch.split(up_proj, row_split_dim, dim = 0)
+        split_down_tensors = torch.split(down_proj, col_split_dim, dim = -1)
+        for i in range(len(split_gate_tensors)):
+            if dist.get_rank()==i:
+                layer.mlp.gate_proj.weight.data = split_gate_tensors[i].to(f"{device}:{i}")
+                layer.mlp.up_proj.weight.data = split_up_tensors[i].to(f"{device}:{i}")
+                layer.mlp.down_proj.weight.data = split_down_tensors[i].to(f"{device}:{i}")
+   
+        bind(layer.mlp, mlp_forward, "forward") 
+    
+# bind the new forward function with llama2 model's self_attn
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+
+    cos = cos.unsqueeze(unsqueeze_dim)
+    sin = sin.unsqueeze(unsqueeze_dim)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+    """
+    This is the equivalent of torch.repeat_interleave(x, dim=1, repeats=n_rep). The hidden states go from (batch,
+    num_key_value_heads, seqlen, head_dim) to (batch, num_attention_heads, seqlen, head_dim)
+    """
+    batch, num_key_value_heads, slen, head_dim = hidden_states.shape
+    if n_rep == 1:
+        return hidden_states
+    hidden_states = hidden_states[:, :, None, :, :].expand(batch, num_key_value_heads, n_rep, slen, head_dim)
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
+    
+def forward(
```

Comment:
In this file, I see two functions `forward` and `mlp_forward`, can you make the naming consistent for better readability?